# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

In [6]:
# import image module 
from IPython.display import Image 

# get the image 
Image(url=r"C:\Users\sombe\Downloads\Iron Kaggle and dl lab-challenges\DL lab-challenges\tttboard.jpg", width=300, height=300) 


Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [15]:
# Import and install important libraries

import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import load_model
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [4]:
# your code here

# Read tic-tac-toe.csv into a dataframe.
df = pd.read_csv(r"C:\Users\sombe\Downloads\Iron Kaggle and dl lab-challenges\DL lab-challenges\tic-tac-toe.csv")

# Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
print(df.head())  # Displays the first 10 rows


# Inspect class column 
unique_values = df['class'].unique()
print(unique_values)
print(df['class'].value_counts())
print(df['class'].dtype)


# # Convert the categorical values to numeric in all columns.
mapping = {'x': 1, 'o': -1, 'b': 0}

for col in ['TL', 'TM', 'TR', 'ML', 'MM', 'MR', 'BL', 'BM', 'BR']:
    df[col] = df[col].map(mapping)

df['class'] = df['class'].astype(int)  # Overwrites 'class' with numerical values

# Inspect dataset
print(df.head())  # Displays the first 10 rows

# Separate the inputs and output.

# Define input features (board positions)
X = df[['TL', 'TM', 'TR', 'ML', 'MM', 'MR', 'BL', 'BM', 'BR']]

# Define output (game outcome)
y = df['class']

# Normalize the input data.

# Initialize the scaler
scaler = StandardScaler()

# Fit and transform the input data (X)
X_normalized = scaler.fit_transform(X)

# Convert normalized data back to a DataFrame if needed
X_normalized = pd.DataFrame(X_normalized, columns=X.columns)

# Display the first few rows
print(X_normalized.head())







  TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True
[ True False]
class
True     626
False    332
Name: count, dtype: int64
bool
   TL  TM  TR  ML  MM  MR  BL  BM  BR  class
0   1   1   1   1  -1  -1   1  -1  -1      1
1   1   1   1   1  -1  -1  -1   1  -1      1
2   1   1   1   1  -1  -1  -1  -1   1      1
3   1   1   1   1  -1  -1  -1   0   0      1
4   1   1   1   1  -1  -1   0  -1   0      1
        TL       TM       TR       ML        MM        MR        BL        BM  \
0  1.03517  1.10683  1.03517  1.10683 -1.241994 -1.223594  1.035170 -1.223594   
1  1.03517  1.10683  1.03517  1.10683 -1.241994 -1.223594 -1.231556  1.106830   
2  1.03517  1.10683  1.03517  1.10683 -1.241994 -1.223594 -1.231556 -1.223594   
3  1.03517  1.10683  1.03517  1.10683 -1.241994 -1.223594 -1.231556 -0.058382   
4  1.03517  1.10683  1.03517  1.

## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [9]:
# your code here

# 1. Split the training and test data.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 2. Create a Sequential model
model = Sequential()

# 3. Add layers to the model
# Input layer: match the number of features in your dataset
model.add(Dense(64, input_shape=(X_train.shape[1],), activation='relu'))  # First hidden layer
model.add(Dense(32, activation='relu'))  # Second hidden layer
model.add(Dense(16, activation='relu'))  # Third hidden layer
model.add(Dense(2, activation='softmax'))  # Output layer with 2 classes (softmax for probabilities)

# 4. Compile the model
model.compile(optimizer='adam', 
              loss='sparse_categorical_crossentropy', 
              metrics=['accuracy'])

# 5. Fit the training data
model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2)

# 6. Evaluate the model with the test data
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Test Loss: {test_loss}, Test Accuracy: {test_accuracy}")

# 7. Save the model as 'tic-tac-toe.model'
model.save('tic-tac-toe_model.keras')
print("Model saved as 'tic-tac-toe_model.keras'")

C:\Users\sombe\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 6s 54ms/step - accuracy: 0.4321 - loss: 0.7490 - val_accuracy: 0.7532 - val_loss: 0.6112
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 20ms/step - accuracy: 0.7252 - loss: 0.6068 - val_accuracy: 0.7727 - val_loss: 0.5479
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.7457 - loss: 0.5520 - val_accuracy: 0.7597 - val_loss: 0.5108
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.7969 - loss: 0.4893 - val_accuracy: 0.7468 - val_loss: 0.4781
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 21ms/step - accuracy: 0.7919 - loss: 0.4702 - val_accuracy: 0.7662 - val_loss: 0.4511
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.8022 - loss: 0.4379 - val_accuracy: 0.7792 - val_loss: 0.4220
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 19ms/step - accuracy: 0.8496 - loss: 0.3811 - val_accuracy: 0.8247 - val_loss: 0.3953
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 20ms/step - accuracy: 0.8687 - loss: 0.3761 - val_accuracy: 0.8247 - v

ValueError: Invalid filepath extension for saving. Please add either a `.keras` extension for the native Keras format (recommended) or a `.h5` extension. Use `model.export(filepath)` if you want to export a SavedModel for use with TFLite/TFServing/etc. Received: filepath=tic-tac-toe.model.

In [13]:
# 7. Save the model as 'tic-tac-toe.model'
model.save('tic-tac-toe_model.keras')
print("Model saved as 'tic-tac-toe_model.keras'")

Model saved as 'tic-tac-toe_model.keras'


## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [21]:
# your code here

# Load the saved model
model = load_model('tic-tac-toe_model.keras')  # or 'tic-tac-toe.h5'

# Select a few random rows from the test set
random_indices = np.random.choice(X_test.shape[0], size=5, replace=False)
X_random = X_test.iloc[random_indices]
y_random = y_test.iloc[random_indices]  # Actual labels for the random rows

# Reset the index of y_random to make sure it's integer-based for easy access
y_random = y_random.reset_index(drop=True)

# Make predictions
predictions = model.predict(X_random)

# Get the predicted class labels (index with highest probability)
predicted_labels = np.argmax(predictions, axis=1)

# Compare predictions with true labels
for i in range(len(predicted_labels)):
    print(f"True label: {y_random[i]}, Predicted label: {predicted_labels[i]}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 269ms/step
True label: 0, Predicted label: 0
True label: 0, Predicted label: 0
True label: 0, Predicted label: 0
True label: 1, Predicted label: 1
True label: 1, Predicted label: 1


In [23]:
# Evaluate the model on the test data
test_loss, test_accuracy = model.evaluate(X_test, y_test)

# Print the results
print(f"Test Loss: {test_loss}")
print(f"Test Accuracy: {test_accuracy}")


6/6 ━━━━━━━━━━━━━━━━━━━━ 1s 20ms/step - accuracy: 0.9798 - loss: 0.0419 
Test Loss: 0.05615432932972908
Test Accuracy: 0.9791666865348816


## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [ ]:
# I did not need to improve the model.

**Which approach(es) did you find helpful to improve your model performance?**

Using the specifications in the instructions given in this challenge, I achieved Test Loss: 0.05615432932972908
Test Accuracy: 0.9791666865348816.
